File: seeed/2-image-classifier/modules/image-classifier-csharp/ImageClassifier.cs
namespace ImageClassifierCsharp
{
    using System;
    using System.Collections.Generic;
    using System.IO;
    using System.Linq;
    using TensorFlow;

    /// <summary>
    /// Tensorflow image classifier.  Input is an image (as jpg encoded bytes) and
    /// runs it thorugh a Tensorflow image classification model to predict labels.
    /// </summary>
    public class ImageClassifier
    {
        private readonly TFGraph _graph;
        private readonly Dictionary<int, string> _labelMap;
        private readonly string _inputName;
        private readonly string _outputName;
        private readonly int _width;
        private readonly int _height;
        private readonly float _mean;
        private readonly float _scale;

        public ImageClassifier(
            string modelFile,
            string labelFile,
            string metadataFile,
            string input,
            string output,
            int width,
            int height,
            float mean,
            float scale)
        {
            // Load Tensorflow model.
            _graph = new TFGraph();
            _graph.Import(File.ReadAllBytes(modelFile));

            // Load labels.
            _labelMap = CreateLabelMap(labelFile, metadataFile);

            // Model parameters.
            _inputName = input ?? throw new ArgumentNullException(nameof(input));
            _outputName = output ?? throw new ArgumentNullException(nameof(output));
            _width = width;
            _height = height;
            _mean = mean;
            _scale = scale;
        }

        /// <summary>
        /// Classifies the image (jpg encoded bytes).  Returns the top 2 predicted labels along with their prediction confidences.
        /// </summary>
        public ImageClassification[] Classify(string name, byte[] image)
        {
            // Decode the jpg image bytes.
            using (var tensor = CreateTensorFromImage(image))
            using (var session = new TFSession(_graph))
            {
                // Configure Tensorflow session.
                var runner = session.GetRunner()
                    .AddInput(_graph[_inputName][0], tensor)
                    .Fetch(_graph[_outputName][0]);

                // Run classifier.
                var output = runner.Run();
                var result = (float[,])(output[0].GetValue());

                // Return top predicted labels.
                return GetPredictions(name, result);
            }
        }

        private ImageClassification[] GetPredictions(string name, float[,] scores)
        {
            // Join labels with predictions.
            var predictions = new ImageClassification[_labelMap.Count];
            for (int i = 0; i < predictions.Length; i++)
            {
                predictions[i] = new ImageClassification { Label = _labelMap[i], Score = scores[0, i] };
            }

            // Return the top 3 label predictions.
            return predictions.OrderByDescending(p => p.Score).Take(3).ToArray();
        }

        private TFTensor CreateTensorFromImage(byte[] contents, TFDataType dataType = TFDataType.Float)
        {
            // DecodeJpeg uses a scalar String-valued tensor as input.
            using (var tensor = TFTensor.CreateString(contents))
            using (var graph = ConstructGraphToNormalizeImage(out TFOutput input, out TFOutput output, dataType))
            using (var session = new TFSession(graph))
            {
                var normalized = session.Run(
                    inputs: new[] { input },
                    inputValues: new[] { tensor },
                    outputs: new[] { output });

                return normalized[0];
            }
        }

        // The inception model takes as input the image described by a Tensor in a very
        // specific normalized format (a particular image size, shape of the input tensor,
        // normalized pixel values etc.).
        //
        // This function constructs a graph of TensorFlow operations which takes as
        // input a JPEG-encoded string and returns a tensor suitable as input to the
        // inception model.
        private TFGraph ConstructGraphToNormalizeImage(out TFOutput input, out TFOutput output, TFDataType destinationDataType = TFDataType.Float)
        {
            var graph = new TFGraph();
            input = graph.Placeholder(TFDataType.String);

            output = graph.Cast(
                graph.Div(
                    x: graph.Sub(
                        x: graph.ResizeBilinear(
                            images: graph.ExpandDims(
                                input: graph.Cast(
                                    graph.DecodeJpeg(contents: input, channels: 3), DstT: TFDataType.Float),
                                dim: graph.Const(0, "make_batch")),
                            size: graph.Const(new int[] { _width, _height }, "size")),
                        y: graph.Const(_mean, "mean")),
                    y: graph.Const(_scale, "scale")), destinationDataType);

            return graph;
        }

        /// <summary>
        /// Create a dict mapping label id to human readable string.
        /// Returns:
        ///    labels_to_names: dictionary where keys are integers from to 1000
        ///    and values are human-readable names.
        /// We retrieve a synset file, which contains a list of valid synset labels used
        /// by ILSVRC competition. There is one synset one per line, eg.
        ///    #   n01440764
        ///    #   n01443537
        /// We also retrieve a synset_to_human_file, which contains a mapping from synsets
        /// to human-readable names for every synset in Imagenet. These are stored in a
        /// tsv format, as follows:
        ///    #   n02119247    black fox
        ///    #   n02119359    silver fox
        /// We assign each synset (in alphabetical order) an integer, starting from 1
        /// (since 0 is reserved for the background class).
        /// Code is based on
        /// https://github.com/tensorflow/models/blob/master/research/inception/inception/data/build_imagenet_data.py#L463
        /// </summary>
        private Dictionary<int, string> CreateLabelMap(string labelFile, string metadataFile)
        {
            var synset_list = File.ReadAllLines(labelFile);
            var synset_to_human_list = File.ReadAllLines(metadataFile);

            var synset_to_human = new Dictionary<string, string>();
            foreach (var s in synset_to_human_list)
            {
                var parts = s.Trim().Split('\t');
                var synset = parts[0];
                var human = parts[1];
                synset_to_human[synset] = human;
            }

            int label_index = 1;
            var labels_to_names = new Dictionary<int, string> { { 0, "background" } };
            foreach (var synset in synset_list)
            {
                var name = synset_to_human[synset];
                labels_to_names[label_index] = name;
                label_index++;
            }

            return labels_to_names;
        }
    }
}

File: seeed/2-image-classifier/modules/image-classifier-python/imageclassifier.py
import numpy as np
from PIL import Image
import tensorflow as tf


class ImageClassifier(object):

    def __init__(self, FLAGS):
        self.FLAGS = FLAGS

        # Creates graph from saved GraphDef.
        self.create_graph()
        # Creates node id --> English label lookup.
        self.label_map = self.create_label_map()

    def run_inference_on_image(self, image):
        """
        Runs image recognition on a .jpg image.  The image is received in a
        Flask http request, converted to a numpy array, reshaped to the required
        input shape for the Tensorflow MobileNet model, and run through the model.

        The output of the model is an array of prediction confidences for each
        class the model was trained against. The top K predictions are converted
        to human-readable labels with their prediction scores.
        """
        image_data = self.get_image_data(image)

        with tf.Session() as sess:
            output_tensor = sess.graph.get_tensor_by_name('MobilenetV2/Predictions/Reshape_1:0')
            predictions = sess.run(output_tensor, {'input:0': image_data})
            predictions = np.squeeze(predictions)

            top_k = predictions.argsort()[-self.FLAGS.num_top_predictions:][::-1]

            results = []
            for node_id in top_k:
                prediction = {
                    'label': self.label_map[node_id],
                    'score': predictions[node_id].astype(float),
                }
                results.append(prediction)

            return results

    def get_image_data(self, image):
        """
        Converts an image from a Flask http request to a numpy array in the
        shape used by the Tensorflow MobileNet model.
        """
        img = np.array(Image.open(image).resize((224,224))).astype(np.float) / 128 - 1
        return img.reshape(1,224,224,3)

    def create_graph(self):
        """
        Creates a graph from frozen GraphDef file.
        """
        with tf.gfile.FastGFile(self.FLAGS.model_path, 'rb') as f:
            graph_def = tf.GraphDef()
            graph_def.ParseFromString(f.read())
            _ = tf.import_graph_def(graph_def, name='')

    def create_label_map(self):
        """
        Create a dict mapping label id to human readable string.
        Returns:
            labels_to_names: dictionary where keys are integers from to 1000
            and values are human-readable names.
        We retrieve a synset file, which contains a list of valid synset labels used
        by ILSVRC competition. There is one synset one per line, eg.
                #   n01440764
                #   n01443537
        We also retrieve a synset_to_human_file, which contains a mapping from synsets
        to human-readable names for every synset in Imagenet. These are stored in a
        tsv format, as follows:
                #   n02119247    black fox
                #   n02119359    silver fox
        We assign each synset (in alphabetical order) an integer, starting from 1
        (since 0 is reserved for the background class).
        Code is based on
        https://github.com/tensorflow/models/blob/master/research/inception/inception/data/build_imagenet_data.py#L463
        """

        synset_list = [s.strip() for s in open(self.FLAGS.label_path).readlines()]
        synset_to_human_list = open(self.FLAGS.label_metadata_path).readlines()

        synset_to_human = {}
        for s in synset_to_human_list:
            parts = s.strip().split('\t')
            assert len(parts) == 2
            synset = parts[0]
            human = parts[1]
            synset_to_human[synset] = human

        label_index = 1
        labels_to_names = {0: 'background'}
        for synset in synset_list:
            name = synset_to_human[synset]
            labels_to_names[label_index] = name
            label_index += 1

        return labels_to_names
